### ISSUE TITLE
Add a framework

### ISSUE THREAD

**Issue by luchaos** [RetroAchievements/RAWeb#42]
## Goals:

- apply php development best practices (avoid globals, pass by reference, echoing out complete html blocks), making code testable
- preserve existing features and api to satisfy external consumers (toolkits, emulators, editor tools)
- DRY (don't repeat yourself) - fixing issues and implementing new features _once_
- streamlined setup for local development to make it attractive for contributors
- SoC (separation of concerns), encapsulate features into components
- simplify and stabilise CRUD (create, read, update, delete) tasks
- principle of least exposure, clean api (no more `.php` in urls)
- avoid huge files (dynrender has multiple thousand LoC) 
- environment agnostic configuration
- global input sanitisation
- common REST/json api
- cli tooling

## TODOs

- [x] Route everything through Laravel with catchall routes -> dispatch static files with legacy bootstrapping
- [x] Adjust .htaccess to single entrypoint, move all rewrite rules to routes (makes it work on nginx as well, no more dependency on apache)
- [x] Adjust .htaccess to not rely on images host apache redirects -> static host config & static_asset() view helper for aws hosted images
- [x] Auth service provider (reuse existing auth cookie, support salted md5 hashes through legacy user provider) to be replaced with https://github.com/RetroAchievements/RAWeb/issues/99
- [x] lowercase & legacy route redirects in `.htaccess` for apache on prod (to be sure links are case insensitive)
- [x] Replace all `IsAtHome()`, `AT_HOST` and `AT_HOST_DOT` with agnostic environment checks
- [x] Implement generalized grid paginator to be reused among resources' indexviews
- [x] Use service container instead of `global`s -> AppServiceProvider provides `RA\RA`
- [x] decorate resource routes with slugs (game, achievement, news, forum)
- [x] replace all references to service credentials with env vars and configs`
- [x] add layout & views for general error pages (404, 500, 503)
- [x] replace header location redirects with redirect responses
- [x] set up new server with upgraded runtimes
- [x] move `request*.php` to controller actions
- [x] email address to .env for error reporting
- [x] Custom logging (replace `error_log()`)
- [x] environment check for recaptcha
- [x] provide a deployment script/tool
- [x] Flash errorCodes to session
- [x] use SES for email sending
- [x] Global exception handling
- [x] Set up frontend pipeline
- [x] add csrf token to all non-api post request forms
- [ ] Move facebook fields to user_connections
- [ ] AWS S3 uploads with test bucket
- [ ] Password Resets
- [ ] Mail authoring

### Tests:

- [ ] Integrations: make sure all app api v1 endpoints (`/*.php`) are working for external consumers. see https://github.com/RetroAchievements/RAWeb/issues/90 and https://github.com/RetroAchievements/RAWeb/issues/91 for details
- [ ] test all users are found regardless of casing (case insensitive)
- [ ] test all page responses for server errors
- [ ] test all api v1 (`/API/*.php`) responses
- [ ] test all legacy routes are redirected
- [ ] test all rss feed responses

### Database:

To ease collaboration Laravel's database migration and seeding features are used (via `artisan` cli tool).
Allows to document and roll out changes made to the database structure for all contributors.

Laravel provides those features out of the box.

- [x] Set up initial database migration
- [x] Provide mysqli support (legacy) -> AppServiceProvider provides `RA\RA::mysqliConnection()`
- [x] Set up models and their relationships using Eloquent (initally for seeds only)
- [x] Migrate database to charset `utf8mb4`, collate `utf8mb4_unicode_ci`
- [x] Migrate columns of type [BIT to boolean/tinyint](https://stackoverflow.com/questions/36225328/laravel-give-data-type-bit-byte-in-migration)
- [x] Set up factories and database seeds
- [x] Migrate table and column names to meet framework conventions
- [x] Re-establish missing foreign keys
- [x] Email Confirmations -> verification field on users
- [x] Add Emulators
- [x] Comment Types to polymorphic relationship
- [ ] Laravel Passport with implicit token grant for the frontend
- [ ] Materialize User <-> Game relations
- [ ] Support Markdown
- [ ] Add Awards
- [ ] Add Events (AotW)

### Seeds:

- [x] Achievements
- [x] Consoles
- [x] Forums
- [x] Forum Categories
- [x] Forum Topics
- [x] Games
- [x] Users
- [x] News

Ignored for now:
- [ ] Achievement Tickets
- [ ] Awards (polymorphic)
- [ ] Comments (polymorphic: User Wall, Forum Topic)
- [ ] Events (polymorphic)
- [ ] Game Alternatives
- [ ] Game Code Notes
- [ ] Game Hashes
- [ ] Leaderboards
- [ ] Leaderboard Entries
- [ ] Messages
- [ ] Ratings (polymorphic: Games, ...)
- [ ] Streams
- [ ] User Achievements (Awarded)
- [ ] User Activities
- [ ] User Awards
- [ ] User History
- [ ] User Relations (Friends, Blocks)
- [ ] Votes (polymorphic: Achievements, ...)

**Comment by luchaos** [comment]
**some seeds in place**

![screen shot 2018-03-14 at 22 52 57](https://user-images.githubusercontent.com/1280590/37433023-a79ae172-27da-11e8-8311-272791a5c623.png)

**default users for all roles**

<img width="496" alt="screen shot 2018-03-14 at 22 56 36" src="https://user-images.githubusercontent.com/1280590/37433125-fae62b7a-27da-11e8-9d69-d14128c0f50d.png">

**error pages with base template**

![screen shot 2018-03-14 at 22 52 43](https://user-images.githubusercontent.com/1280590/37433029-aa1c8aa4-27da-11e8-972c-56372fd43646.png)

**maintenance mode**

after calling `php artisan down --message "maintenance"` on command line the appication answers all requests with http status 503 (both we and api requests)
allows for larger database migrations without interruptions

![screen shot 2018-03-14 at 22 43 36](https://user-images.githubusercontent.com/1280590/37433039-b12ef412-27da-11e8-80b2-c872f76d456b.png)
![screen shot 2018-03-14 at 22 51 54](https://user-images.githubusercontent.com/1280590/37433033-acebf166-27da-11e8-909e-1809f4e80076.png)
![screen shot 2018-03-14 at 22 44 01](https://user-images.githubusercontent.com/1280590/37433035-af18100a-27da-11e8-97fc-8e04a0621fe8.png)

**Comment by SyrianBallaS** [comment]
The seeding definitely works, that saved a lot of tedious pain.

**Comment by luchaos** [comment]
@SyrianBallaS i'm still in the middle of refactoring - branching out from master is definitely the way to go for now. using this branch just for seeding is a good approach until then.

**Comment by meleu** [comment]
I may be wrong, but looks the seeder created a table named `users` where it should be named `UserAccounts`.

After running `php artisan db:seed`, look what I get from a `show tables;` on the database server:

```
MariaDB [RAWeb]> show tables;
+-------------------------------+
| Tables_in_RAWeb               |
+-------------------------------+
| Achievements                  |
| Activity                      |
| ArticleTypeDimension          |
| Awarded                       |
| Chat                          |
| CodeNotes                     |
| Comment                       |
| Console                       |
| EmailConfirmations            |
| Forum                         |
| ForumCategory                 |
| ForumTopic                    |
| ForumTopicComment             |
| Friends                       |
| GameAlternatives              |
| GameData                      |
| GameHashLibrary               |
| LeaderboardDef                |
| LeaderboardEntry              |
| Messages                      |
| News                          |
| PlaylistVideo                 |
| Rating                        |
| ScoreHistory                  |
| SiteAwards                    |
| StaticData                    |
| Ticket                        |
| Votes                         |
| migrations                    |
| oauth_access_tokens           |
| oauth_auth_codes              |
| oauth_clients                 |
| oauth_personal_access_clients |
| oauth_refresh_tokens          |
| password_resets               |
| users                         |
+-------------------------------+
36 rows in set (0.00 sec)
```

I think that last table named `users` should be `UserAccounts`.

**Comment by luchaos** [comment]
that's intended - more tables will be renamed to match the framework naming convention in the upcoming updates.
https://github.com/RetroAchievements/RAWeb/blob/framework/database/migrations/2018_03_27_100004_apply_users_convention.php#L16

**Comment by meleu** [comment]
(decided to post here what we were talking on discord)
@luchaos while working on RAWeb's code in the last few weeks I noticed the lack of some info that IMHO should be currently available.

Probably most, if not all, of the fields listed here are already addressed by other issues and/or by what you have in mind for V2. But I'll list them here just to assure that we don't forget them. ;)

**UserAccounts**: `MemberSince` (timestamp)

**Achievements**: `ModifiedByUserID` (int) - maybe unecessary if you're going to use another approach for the audit log.

**LeaderboardDef**: `Author` (varchar)

**Comment by linusonline** [comment]
Have you considered migrating the frontend to a modern web framework, like react.js? If/when there is a JSON based REST API, this task should be completely separable from the backend code, and would not need to change it. (I'm not familiar with the current site architecture, but speaking from experience with both frameworks, migrating a PHP/jQuery based frontend to react.js seems like it should make further development *a lot* easier, faster and more powerful.)

**Comment by luchaos** [comment]
@linusonline absolutely, thanks for the suggestion!
the current state of things is a very basic "flat" php project with jquery sprinkled in - both very dated in its implementation.
so, for a roadmap, the current plan is to refactor everything to a laravel application, mainly to improve the overall backend setup but also to have proper means to build more solid apis in the future (rest, json rpc, graph, protobuffed, etc...). this has already been started over a year ago in a private repository and not only adds a lot of new features but also has full legacy support for v1 built in.
as for the frontend, the first step is to go the shortest route for an improved setup to keep the initial workload for a v2 release sensible. it primarily uses laravel's blade templates and some vue.js components to add interactivity; with the goal to replace pretty much all of the current jquery code eventually and making the frontend modern and maintainable (basic responsiveness with bootstrap 4, standard components, webpack pipeline).
after that has been done and v2 is released there will be "industry standards favoring" setups for both backend and frontend - so it will be fairly easy to transition to a full-blown frontend framework, be it vue.js, react or angular is up for discussion, much as the decision to build a SPA around it or stick with server side rendered blade templates in the end.

TL;DR: yes :)

**Comment by linusonline** [comment]
Thanks for the quick and detailed answer! :) Sounds very exciting. I'll be very interested to see how things develop, and maybe even get involved later, if I have the time. My PHP is too rusty to be useful, but if things get into node.js territory, it would be awesome to contribute. Also have some experience with OAS 3.0 and REST API:s. Any particular plan for API documentation? I'm a fan of the OAS generate-from-spec model, enabling things like this to be auto-generated: https://petstore.swagger.io.

**Comment by juherr** [comment]
The latest message on the ticket is one year old.
I didn't found any v2 branch and the latest commits are all based on the current version.

What is the status of the v2?

**Comment by luchaos** [comment]
> The latest message on the ticket is one year old.
> I didn't found any v2 branch and the latest commits are all based on the current version.
> 
> What is the status of the v2?

It is still in development in a different, private repository - used to be in here but was removed as it has nothing in common with the current codebase anymore. 
The application will be split into multiple parts which will be published in new repositories. 
I will update this issue accordingly when it's ready.

**Comment by juherr** [comment]
@luchaos Ok. So if I understand well, it will be useless to contribute on v1 but it is not yet possible to have a look and help at the v2.
Do you have any roadmap?
